wgsl/index.bs

```diff
@@ -3959,6 +3959,278 @@ TODO: deduplicate these tables
 
 ## Texture built-in functions ## {#texture-builtin-functions}
 
+### Sample ### {#texture-sample-functions}
+
+Sampling from a texture uses the `textureSample` function:
+
+```
+<result> textureSample(<texture>, <sampler>, <coordinates> [, <named-arguments>])
+```
+
+Where:
+
+* `<result>` is the texture's parameterized type when using [sampled](#sampled-texture-type)
+    or [read-only](#texture-ro) texture formats.<br>
+* `<result>` is `f32` when using [depth](#texture-depth) texture types.
+
+* `<texture>` is one of the [sampled](#sampled-texture-type),
+    [read-only](#texture-ro) or [depth](#texture-depth) texture types.
+
+* `<sampler>` is a [sampler type](#sampler-type).
+
+* `<coordinates>` are the texture coordinates used for sampling.
+    The type for the texture coordinates depends on the `<texture>` type and is
+    either a `f32`, `vec2<f32>`, `vec3<f32>` or `vec4<f32>`:
+
+    <table class='data'>
+      <thead>
+        <tr><td>`<texture>` type<td>`<coordinates>` type
+      </thead>
+      <tr><td> `texture_sampled_1d<type>`         <td> `f32`
+      <tr><td> `texture_sampled_1d_array<type>`   <td> `vec2<f32>`
+      <tr><td> `texture_sampled_2d<type>`         <td> `vec2<f32>`
+      <tr><td> `texture_sampled_2d_array<type>`   <td> `vec3<f32>`
+      <tr><td> `texture_sampled_3d<type>`         <td> `vec3<f32>`
+      <tr><td> `texture_sampled_cube<type>`       <td> `vec3<f32>`
+      <tr><td> `texture_sampled_cube_array<type>` <td> `vec4<f32>`
+      <tr><td> `texture_ro_1d<type>`              <td> `f32`
+      <tr><td> `texture_ro_1d_array<type>`        <td> `vec2<f32>`
+      <tr><td> `texture_ro_2d<type>`              <td> `vec2<f32>`
+      <tr><td> `texture_ro_2d_array<type>`        <td> `vec3<f32>`
+      <tr><td> `texture_ro_3d<type>`              <td> `vec3<f32>`
+      <tr><td> `texture_depth_2d`                 <td> `vec2<f32>`
+      <tr><td> `texture_depth_2d_array`           <td> `vec3<f32>`
+      <tr><td> `texture_depth_cube`               <td> `vec3<f32>`
+      <tr><td> `texture_depth_cube_array`         <td> `vec4<f32>`
+    </table>
+
+    Array textures (with `_array` suffix) use the last component of `<coordinates>`
+    to specify array index.
+
+
+<div class='example' heading="Sample">
+  <xmp>
+    color = textureSample(tex2D, s, coords);
+  </xmp>
+</div>
+
+Following the `<coordinates>` parameter, a number of additional named arguments may
+be passed for the sample operation:
+
+#### Sample with texel offset #### {#texture-sample-with-offset}
+
+```
+<result> textureSample(<texture>, <sampler>, <coordinates>, offset: <offset> [, <named-arguments>])
```

Comment:
optional arguments need to be discussed with the group first. I think it could be an unnecessary complication for the language.

Reply:
Ack. Alternative suggestions are welcome.

I just wanted to avoid a whole bunch of `textureSampleXXX()` flavours, or overloads which might become ambiguous as more options are added over time.